### ISSUE TITLE
Are local cached files checked against the contents of conanmanifest.txt?

### ISSUE THREAD

**Issue by mbitsnbites** [conan-io/conan#3821]
Seeing that all information about the contents of a conan package is given in `conanmanifest.txt`, is it ever used to check that the locally cached files are correct?

I would expect that running `conan install ...` on an already installed package would make sure that the local copy is intact, or either complain or re-download the package.

However, after manually corrupting a file in the cache and then running install, all I got was `Already installed!`.

Is there some way to explicitly do an integrity check of a specific package?

**Comment by memsharded** [comment]
There was some checks, but they were very slow and a few users asked to remove them, so they were removed. Now, they remain as an option of upload, and could be checked with:

```bash
$ conan upload zlib/1.2.11@conan/stable --skip-upload --check
```

Maybe this should be an option too for ``conan install``, lets discuss it.

**Comment by mbitsnbites** [comment]
We want to use conan in our CI build system, which is a quite busy system. From experience all things that can happen, will happen (e.g. network outage, full disks, faulty RAM:s, buggy SATA controllers, build software bugs, etc, etc), and tracking down and mending subtle errors after the fact is a pain.

Thus it's important to be able to tell that the build environment is valid before starting a job, and hence a check during/after conan install would be helpful. For our purposes, an explicit "check" command would be OK, but it could just as well be an opt-in argument to conan install.

If it's too slow, I think that it should be possible to make a faster check based on file sizes & timestamps (similar to rsync), that should suffice in most cases (size difference => check fail, timestamp difference => proceed to checksum check). It's also possible that making use of all CPU cores (e.g. using a ThreadPool) could help, if that's not already the case.

**Comment by puetzk** [comment]
It's also easier than I'd really prefer to click on a header file in the IDE (e.g. go to error), notice a bug, fix it, and not realize you just edited the packaged copy instead of the source (especially if you commonly use workspaces, where that *is* a valid thing to do). Which leads right back to the "it works on my machine!" issues packaging was supposed to fix :-)

I don't disagree with the current behavior being the default, but a conan install --check (or even just conan fsck <reference>) would be nice-to-have, as would a way to mark files in the cache read-only.

**Comment by puetzk** [comment]
> as would a way to mark files in the cache read-only.

Nevermind, that one apparently exists: [CONAN_READ_ONLY_CACHE](https://docs.conan.io/en/latest/reference/env_vars.html#conan-read-only-cache).

**Comment by memsharded** [comment]
I am having a look at this issue, and analyzing the possibility of such ``--check`` argument.

As a side note, I wanted to point out @mbitsnbites and @puetzk that package installation has guards. If something fails (network, unzipping, disk full...) and the operations don't complete successfully, the packages are automatically marked as dirty, and subsequent commands will remove them and install them again. Please report if you don't see this behavior when something fails.

**Comment by memsharded** [comment]
This makes sense and is doable. A few implementation notes:

- Extract the functionality of checking the integrity of packages to a common place (now in RemoteManager, being used only for ``upload``)
- Package checks should be done also for built, cached, and downloaded packages.
- Argument ``--check`` or the like, can be common to install, create, download, and upload.

**Comment by mbitsnbites** [comment]
@memsharded that sounds excellent!

**Comment by 0x8000-0000** [comment]
Ping? We have encountered mysterious errors on one system which did not occur on others. The cache is marked read-only everywhere, but still some files were missing.

A simple tool that could be triggered to check a package or check all would be helpful here.

**Comment by SK83RJOSH** [comment]
Another year, another bump! We're running into issues with incomplete packages en masse at work and we've had to resort to playing wack-a-mole by removing packages that fail to build and hope they get pulled down correctly this time around. We've not identified any particular cause for the failures in the first place, but having a tool to validate them before we kick off and wait for a lengthy build would be really great. ❤️

**Comment by memsharded** [comment]
Hi, 

You might try the ``--check`` functionality in the ``upload`` command, that validates packages contents against their manifests. something like:
```bash
conan upload * -c --skip-upload --check
```
could help

**Comment by vajdaz** [comment]
> Hi,
> 
> You might try the `--check` functionality in the `upload` command, that validates packages contents against their manifests. something like:
> 
> ```shell
> conan upload * -c --skip-upload --check
> ```
> 
> could help

It helps. However, my use case would be to do this kind of check after `conan install` (maybe nothing had to be downloaded, package found in cache, I wonder if cache contents are still ok). So maybe `conan install` should have a `--check` option which would make the exact same check at the end of the installation command as `conan upload` does. It may even work if the remote is not rachable (`conan upload` requires an accessible remote, even with `--skip-upload`).

What do you think about the idea op adding a checking option to the installation command?

**Comment by memsharded** [comment]
Yes, this would make sense. The only thing is that this won't fit in the 1.X scope anymore, and it is super unlikely this will be prioritized into the 2.0, but it will be after it.

**Comment by memsharded** [comment]
Besides the ``conan upload --check``, Conan 2.0 provided a new ``conan cache check-integrity`` command that will do exactly this, to verify the cached files against the manifests.

Closing this as implemented, feedback welcome!
